File: observability/instrumentation/node/business.ts
/**
 * Business metric helpers for Node services (offcon_* series).
 * payment-svc and notification-svc emit these.
 */

import { Counter, Gauge } from "prom-client";
import { registry } from "./observability";

export const businessMetrics = {
  // payment-svc
  paymentAttempts: new Counter({
    name: "offcon_payment_attempts_total",
    help: "Payment attempts.",
    labelNames: ["result"], // succeeded | failed
    registers: [registry],
  }),
  subscriptionMrrCents: new Gauge({
    name: "offcon_subscription_mrr_cents",
    help: "Monthly recurring revenue in cents.",
    registers: [registry],
  }),
  activeSubscriptions: new Gauge({
    name: "offcon_active_subscriptions",
    help: "Active subscriptions by plan.",
    labelNames: ["plan"],
    registers: [registry],
  }),

  // notification-svc
  notificationQueueDepth: new Gauge({
    name: "offcon_notification_queue_depth",
    help: "Pending notifications in the delivery queue.",
    registers: [registry],
  }),
  notificationsSent: new Counter({
    name: "offcon_notifications_sent_total",
    help: "Notifications delivered.",
    labelNames: ["channel", "result"], // channel=email|push|in_app, result=success|failed
    registers: [registry],
  }),
};

File: observability/instrumentation/go/business.go
package observability

import "github.com/prometheus/client_golang/prometheus"

// BusinessMetrics exposes the offcon_* custom metrics that power the business
// KPI dashboard + alerts. Services register only the ones they emit.
//
// Example (orchestrator):
//
//	bm := observability.RegisterBusinessMetrics()
//	bm.InstanceSpawns.WithLabelValues("success").Inc()
//	bm.InstancesActive.Set(float64(running))
type BusinessMetrics struct {
	// orchestrator
	InstanceSpawns  *prometheus.CounterVec // result=success|error
	InstancesActive prometheus.Gauge
	// flag-verifier / scoring
	FlagSubmissions *prometheus.CounterVec // result=accepted|rejected
	FlagVerifyErr   prometheus.Counter
	CtfSolves       *prometheus.CounterVec // event_id
	// auth / user
	Signups prometheus.Counter
	// payment
	PaymentAttempts *prometheus.CounterVec // result=succeeded|failed
	SubscriptionMRR prometheus.Gauge
	ActiveSubs      *prometheus.GaugeVec // plan
	// notification
	NotificationQueue prometheus.Gauge
}

// RegisterBusinessMetrics registers all business collectors against the
// service's metrics registry. Unused metrics simply stay at zero.
func RegisterBusinessMetrics() *BusinessMetrics {
	reg := globalMetrics.registry

	bm := &BusinessMetrics{
		InstanceSpawns: prometheus.NewCounterVec(prometheus.CounterOpts{
			Name: "offcon_instance_spawn_total", Help: "Lab instance spawn attempts.",
		}, []string{"result"}),
		InstancesActive: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "offcon_instances_active", Help: "Currently running lab instances.",
		}),
		FlagSubmissions: prometheus.NewCounterVec(prometheus.CounterOpts{
			Name: "offcon_flag_submissions_total", Help: "Flag submissions.",
		}, []string{"result"}),
		FlagVerifyErr: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "offcon_flag_verify_errors_total", Help: "Flag verification internal errors.",
		}),
		CtfSolves: prometheus.NewCounterVec(prometheus.CounterOpts{
			Name: "offcon_ctf_solves_total", Help: "CTF challenge solves.",
		}, []string{"event_id"}),
		Signups: prometheus.NewCounter(prometheus.CounterOpts{
			Name: "offcon_signups_total", Help: "New user signups.",
		}),
		PaymentAttempts: prometheus.NewCounterVec(prometheus.CounterOpts{
			Name: "offcon_payment_attempts_total", Help: "Payment attempts.",
		}, []string{"result"}),
		SubscriptionMRR: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "offcon_subscription_mrr_cents", Help: "Monthly recurring revenue in cents.",
		}),
		ActiveSubs: prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Name: "offcon_active_subscriptions", Help: "Active subscriptions by plan.",
		}, []string{"plan"}),
		NotificationQueue: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "offcon_notification_queue_depth", Help: "Pending notifications in the delivery queue.",
		}),
	}

	reg.MustRegister(
		bm.InstanceSpawns, bm.InstancesActive, bm.FlagSubmissions, bm.FlagVerifyErr,
		bm.CtfSolves, bm.Signups, bm.PaymentAttempts, bm.SubscriptionMRR,
		bm.ActiveSubs, bm.NotificationQueue,
	)
	return bm
}
